### PR TITLE
3.2 function returntype

### PR DESCRIPTION
**Pull request by lorenzo**
closes https://github.com/cakephp/cakephp/issues/4283

### CODE REVIEW

File: src/Database/FunctionsBuilder.php [hunk 1]
```diff
@@ -31,11 +31,12 @@ class FunctionsBuilder
      * @param string $name the name of the SQL function to constructed
      * @param array $params list of params to be passed to the function
      * @param array $types list of types for each function param
+     * @param string $return The return type of the function expression
      * @return FunctionExpression
      */
-    protected function _build($name, $params = [], $types = [])
+    protected function _build($name, $params = [], $types = [], $return = 'string')
```
Comment (dereuromark) [inline src/Database/FunctionsBuilder.php]: Can we make this steing a class constant? Better to use then from the outside and cleaner as magic strings then IMO.

File: tests/TestCase/ORM/QueryRegressionTest.php [hunk 2]
```diff
@@ -963,13 +963,31 @@ public function testTypemapInFunctions()
...
+     *
+     * @return void
+     */
+    public function testTypemapInFunctions2()
+    {
+        $table = TableRegistry::get('Comments');
+        $query = $table->find();
+        $query->select([
+            'id',
+            'max' => $query->func()->max('created', ['datetime'])
```
Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: Why is the last argument an array?

**Comment by markstory** [comment]
Awesome work @lorenzo. Nice that it wasn't that invasive once the refactoring around the caster was in place.

File: tests/TestCase/ORM/QueryRegressionTest.php (hunk 2 above)
Comment (lorenzo) [inline tests/TestCase/ORM/QueryRegressionTest.php]: The types argument is always an array

File: src/Database/FunctionsBuilder.php (hunk 1 above)
Comment (lorenzo) [inline src/Database/FunctionsBuilder.php]: maybe... We already use the magic strings everywhere, they are as magic as PHP type names

Comment (dereuromark) [inline src/Database/FunctionsBuilder.php]: :trollface:

File: tests/TestCase/ORM/QueryRegressionTest.php (hunk 2 above)
Comment (markstory) [inline tests/TestCase/ORM/QueryRegressionTest.php]: Ah right.

**Comment by lorenzo** [comment]
Seems like the CS checker is impossible to please :(

File: src/ORM/Query.php [hunk 3]
```diff
@@ -18,6 +18,7 @@
 use Cake\Database\ExpressionInterface;
 use Cake\Database\Query as DatabaseQuery;
 use Cake\Database\TypeMap;
+use Cake\Database\TypedResultInterface;
```
Comment (antograssiot) [inline src/ORM/Query.php]: @Lorenzo it seems that you just need to move this line up and ise alphabetical order to please phpcs.

Comment (lorenzo) [inline src/ORM/Query.php]: But that is not the right alphabetic order if I move it!

Comment (lorenzo) [inline src/ORM/Query.php]: Well it actually is... for some reason vim refuses to correct it for me

**Pull request merged**